### DISCUSSION TITLE
Custom Prompt Properties

### DISCUSSION THREAD

**Discussion by ivannin** [logancyang/obsidian-copilot#794] [Ideas]
Custom Prompts is a very powerful tool, and I, for example, use them almost constantly when working with my notes.

I believe it would be very convenient to set certain modes when Custom Prompts executing.  I currently lack the following functions when using Custom Prompts:

- Switching Copilot mode.
- Switching the model used.
- Note rendering mode when adding it to the context.

I think it would be very convenient to set these modes using YAML properties in the front matter of the note containing the Custom Prompt.

Here are some examples how this could be used:

## Switching Copilot Mode

For example, I'm working in Vault QA mode and want to generate a list of tags for the current note. This is a function I use constantly; it's been assigned to a hotkey.  This is essentially calling Apply Custom Prompt with text describing the tag list generation in the form I need.

This prompt should be executed in chat mode, but I have Vault QA selected, so the prompt executes much slower than I would like.

It would be great to specify something like this in the prompt template:

```
---
Copilot_mode: chat
Copilot_mode_restore: true
---

Generate list of tags for this note {activeNote}.
```

The `Copilot_mode` parameter switches the required mode before execution, and the `Copilot_mode_restore` parameter indicates whether to restore the previous mode after prompt execution or not.

## Switching the Model Used

I use several models. In my daily work, I mainly use the `gpt-4o-mini` and `gemini-flash-1.5` models, but other models are required for certain tasks. For example, for a number of more complex queries, I use `claude-3.5-sonnet`, and for queries related to actual data — `perplexity/llama-3.1-sonar-large-128k-online`.

For example, this is how a Custom Prompt "Weather Forecast" might look:

```
---
Copilot_model: perplexity/llama-3.1-sonar-large-128k-online
Copilot_model_restore: true
---

Show the weather forecast for Moscow for the next week.
```

The `Copilot_model` parameter sets a pre-configured model, and the `Copilot_model_restore` parameter indicates whether to restore the previous value of the selected model after the query is executed.

## Note Rendering Mode

When a note is added to the query context, it is added simply as text. This includes any Dataview blocks or Tasks queries. However, it would be very convenient in some cases to first execute them and add the rendering results to the context.

For example, my daily note template contains this block:

````
```tasks
not done
(due today) OR (scheduled on today) 
sort by priority, created
short mode
```
````

This is how a Custom Prompt "Daily Planning" might look:

```
---
Copilot_render_note: true
---

Based on {activeNote}, plan my day.
```

The `Copilot_render_note` parameter tells Copilot to render all mentioned notes before including them in the query context.

I think these features would significantly expand Copilot's capabilities, and I suggest we discuss them.

**Comment by logancyang** [comment]
So to summarize

1. YAML properties in Custom Prompt for mode, model it should use
2. Execute code blocks such as Dataview before adding to context

These are great suggestions. The second one was mentioned in an FR before and I will prioritize it once Copilot Plus is rolled out.

The first one I need to think about more. With Copilot Plus and the incoming "Profile" feature, there will be no need to switch modes.
